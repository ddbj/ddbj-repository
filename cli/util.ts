import { colorize } from 'json_colorize/mod.ts';
import { colors } from 'cliffy/ansi/colors.ts';

export function requireLogin() {
  console.error(`First you need to log in; run ${colors.bold('`ddbj-repository auth login`')}.`);

  Deno.exit(1);
}

export async function ensureSuccess(res: Response) {
  if (res.ok) return;

  const type = res.headers.get('content-type');

  if (type?.includes('application/json')) {
    const payload = await res.json();

    colorize(JSON.stringify(payload, null, 2));
  } else {
    console.error(`Error: ${res.status} ${await res.text()}`);
  }

  Deno.exit(1);
}
