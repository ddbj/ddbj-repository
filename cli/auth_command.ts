import { delay } from 'std/async/mod.ts';

import * as oauth from 'https://deno.land/x/oauth4webapi@v2.4.0/mod.ts';
import { Command } from 'cliffy/command/mod.ts';
import { colors } from 'cliffy/ansi/colors.ts';
import { open } from 'https://deno.land/x/open@v0.0.6/index.ts';

import { Config, writeConfig } from './config.ts';
import { ensureSuccess } from './util.ts';

export default class extends Command {
  constructor({ issuer, endpoint, apiKey }: Config) {
    super();

    return this
      .action(() => this.showHelp())
      .command('whoami')
      .action(async () => {
        if (apiKey) {
          const uid = await fetchUid(endpoint, apiKey);

          console.log(`Logged in as ${colors.bold(uid)}.`);
        } else {
          console.log('Not logged in.');
        }
      })
      .command('login')
      .action(async () => {
        await openAuthorizationURL(issuer, endpoint);
      })
      .command('logout')
      .action(async () => {
        await writeConfig({ apiKey: undefined });
      })
      .reset();
  }
}

const port = 37376;
const redirectUri = `http://localhost:${port}`;

async function openAuthorizationURL(_issuer: string, endpoint: string) {
  const issuer = new URL(_issuer);

  const res = await oauth.discoveryRequest(issuer);
  const as = await oauth.processDiscoveryResponse(issuer, res);

  const client: oauth.Client = {
    client_id: 'ddbj-repository',
    token_endpoint_auth_method: 'none',
  };

  const state = oauth.generateRandomState();
  const codeVerifier = oauth.generateRandomCodeVerifier();
  const codeChallenge = await oauth.calculatePKCECodeChallenge(codeVerifier);

  const server = Deno.serve(
    {
      port,
      onListen() {
        // do nothing. suppress noisy output
      },
    },
    callbackHandler(as, client, state, codeVerifier, endpoint, async () => {
      await delay(1000);
      await server.shutdown();
    }),
  );

  const authorizationUrl = new URL(as.authorization_endpoint!);

  authorizationUrl.searchParams.set('client_id', client.client_id);
  authorizationUrl.searchParams.set('redirect_uri', redirectUri);
  authorizationUrl.searchParams.set('response_type', 'code');
  authorizationUrl.searchParams.set('scope', 'openid');
  authorizationUrl.searchParams.set('state', state);
  authorizationUrl.searchParams.set('code_challenge', codeChallenge);
  authorizationUrl.searchParams.set('code_challenge_method', 'S256');

  await open(authorizationUrl.href);
}

function callbackHandler(as: oauth.AuthorizationServer, client: oauth.Client, state: string, codeVerifier: string, endpoint: string, done: () => void) {
  return async (req: Request) => {
    const url = new URL(req.url);

    if (url.pathname !== '/') return new Response('Not Found', { status: 404 });

    try {
      const params = oauth.validateAuthResponse(as, client, url, state);

      if (oauth.isOAuth2Error(params)) throw params;

      const res = await oauth.authorizationCodeGrantRequest(as, client, params, redirectUri, codeVerifier);
      const result = await oauth.processAuthorizationCodeOpenIDResponse(as, client, res, oauth.expectNoNonce);

      if (oauth.isOAuth2Error(result)) throw result;

      const apiKey = await obtainAPIKey(endpoint, result.access_token);

      writeConfig({ apiKey });

      const uid = await fetchUid(endpoint, apiKey);

      console.log(`Logged in as ${colors.bold(uid)}.`);

      return new Response(`Logged in as ${uid}.`, { status: 200 });
    } finally {
      done();
    }
  };
}

async function obtainAPIKey(endpoint: string, accessToken: string) {
  const res = await fetch(`${endpoint}/auth/login_by_access_token`, {
    method: 'post',
    headers: {
      'Content-Type': 'application/jwt',
    },
    body: accessToken,
  });

  ensureSuccess(res);

  const { api_key } = await res.json();

  return api_key;
}

async function fetchUid(endpoint: string, apiKey: string) {
  const res = await fetch(`${endpoint}/me`, {
    headers: {
      Authorization: `Bearer ${apiKey}`,
    },
  });

  ensureSuccess(res);

  const { uid } = await res.json();

  return uid;
}
