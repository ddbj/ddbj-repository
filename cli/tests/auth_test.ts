import { returnsNext, stub } from 'std/testing/mock.ts';

import { snapshotTest } from 'cliffy/testing/mod.ts';

import mainCommand from '../main_command.ts';
import { _internals } from '../auth_command.ts';
import { runInContext } from './util.ts';

await snapshotTest({
  name: 'auth whoami',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext({
      responses: [
        new Response(JSON.stringify({ uid: 'alice', admin: false })),
      ],
    }, async () => {
      const readStub = stub(_internals, 'read', returnsNext(['API_KEY']));

      try {
        await mainCommand.parse(['auth', 'whoami']);
      } finally {
        readStub.restore();
      }
    });
  },
});

await snapshotTest({
  name: 'auth whoami; not logged in',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext({}, async () => {
      const readStub = stub(_internals, 'read', returnsNext([undefined]));

      try {
        await mainCommand.parse(['auth', 'whoami']);
      } finally {
        readStub.restore();
      }
    });
  },
});

await snapshotTest({
  name: 'auth login API_KEY',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext({
      responses: [
        new Response(JSON.stringify({ uid: 'alice', admin: false })),
      ],
    }, async () => {
      const writeStub = stub(_internals, 'write', returnsNext([undefined]));

      try {
        await mainCommand.parse(['auth', 'login', 'API_KEY']);
      } finally {
        writeStub.restore();
      }
    });
  },
});

await snapshotTest({
  name: 'auth login; not logged in',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext({
      responses: [
        new Response(JSON.stringify({ login_url: 'http://example.com/login' })),
      ],
    }, async () => {
      // @ts-expect-error for test
      const keypressStub = stub(_internals, 'keypress', returnsNext([Promise.resolve({ key: 'a' })]));
      // @ts-expect-error for test
      const openStub = stub(_internals, 'open', returnsNext([Promise.resolve()]));

      try {
        await mainCommand.parse(['auth', 'login']);
      } finally {
        keypressStub.restore();
        openStub.restore();
      }
    });
  },
});

await snapshotTest({
  name: 'auth logout',
  meta: import.meta,
  denoArgs: ['--allow-all'],

  async fn() {
    await runInContext({}, async () => {
      const removeStub = stub(_internals, 'remove', returnsNext([undefined]));

      try {
        await mainCommand.parse(['auth', 'logout']);
      } finally {
        removeStub.restore();
      }
    });
  },
});
