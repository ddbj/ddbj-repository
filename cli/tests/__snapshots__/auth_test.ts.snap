export const snapshot = {};

snapshot[`auth whoami 1`] = `
stdout:
"Logged in as alice.
"
stderr:
""
`;

snapshot[`auth whoami; not logged in 1`] = `
stdout:
"Not logged in.
"
stderr:
""
`;

snapshot[`auth login API_KEY 1`] = `
stdout:
"Logged in as alice.
"
stderr:
""
`;

snapshot[`auth login 1`] = `
stdout:
"cli: Press any key to open up the browser to login or q to exit:
Opening browser to http://example.com/login

Execute the following command with the API key displayed in your browser:

\$ ddbj-repository auth login YOUR_API_KEY
"
stderr:
""
`;

snapshot[`auth logout 1`] = `
stdout:
""
stderr:
""
`;
